### PR TITLE
Gantt Chart Feature

### LINKED ISSUE #90: Gantt Chart
## Overview
Create a new Page, the projects Gantt chart, thereby supporting the carryover of the google sheets solution.

## Stakeholders
**Product Stakeholder**: @jamescd18, @leoRysing 
**Engineering Stakeholder**: @RChandler234 
**Reference Users**: Jessica Shang, Anushka Wani

## User Stories

As a project manager/lead, I want to be able to have easy access to information regarding a project's timeline, position in relation to other projects and its completion status. By using a Gantt chart, I can easily gleam said information from a diagram.

## Success Metrics

Success will be measured based on the frequency of interaction with this page.

We can record the baseline for this using the Google Activity Dashboard. Namely, we can check the **Viewers** and **Viewer Trends** pages. From there, a simple count is sufficient.

In the app, we can record these using user sessions every time someone accesses the new Gantt page.
- This feature will be successful if we get similar number of unique users over <X_duration>.

**Google Spreadsheet Number of Unique Users:** 6 (based on <# of unique users> from August 2nd to August 17th)
**Google Spreadsheet Daily Avg Users:** 1.62 [round up to 2] (based on <# of users on a given day> averaged from July 27th to August 17th)

Some fun bonus metrics for the dashboard could be adding in timestamps to track the total number of activities to this page.

Activity is as follows;
1.  People navigating to the Gantt chart page
2. People filtering options on the Gantt chart page
3. People clicking from the Gantt chart page

## Background

The current Gantt chart is functional but leaves much to be desired, especially outside of the scope of an individual project. This feature is a high priority as it is a highly-requested feature that greatly. Currently, it only exists as a Spreadsheet solution.

The existing gantt is found at;
https://docs.google.com/spreadsheets/d/1KMaX7BX4c7TXwEU0zSwl87XUk_U6zp7as5WFV6ToyFM/edit#gid=0

## Prerequisites

Prerequisite Tickets;
 - [x] #135 
 - [x] #136 
 
## Acceptance Criteria and Mocks

When created, there should be a new page in the FinishLine menu for the Gantt Chart 

This page should be a large gantt chart. Translate the design of the work package gantt with our existing projects..

Translating the above, the gantt should have the following;
- [x] Each entry should be for a project
- [x] Each project should have its name and its WBS number
  - [x] These top two should work as the label for each project, on the left side of the gantt
- [x] The remainder of the Gantt will be devoted to the timeline calendar
  - [ ] Have a weekly breakdown, with each week starting on the monday
  - [x] The corresponding bar of the project should start at its start date, and stretch proportional to its duration
  - [x] It should use the gantt color feature as a short hand for its percent completion
- [x] When you hover over a project, more info should show up;
  - [x]  You should have its WBS Number and Name on the top of the hover
  - [x] Directly to the right of this, put the start date
  - [x] Add its duration below that
  - [x] Below that, show its duration
  - [x] Below that, show its end date

A mock for this is a follows;
Ignore the filter in the mocks, filters for Gantt is not in scope for this epic.
Screen Shot 2022-08-27 at 6 19 29 PM

Hover should provide more info about these projects;
Screen Shot 2022-08-27 at 6 19 37 PM

If possible, some way to group projects is advisable, may need further investigation.

### Tickets

- [x] #25 
- [x] #286 
- [x] #284 
- [x] #530

### PR DESCRIPTION
**Pull request by Ari-S-123**
## Changes

Gantt Chart Feature 

## Screenshots

Screenshot 2023-01-27 120842

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #90

### CODE REVIEW

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Could we get the sidebar icon to be the second right after home please? Conceptually Gantt is a high-level picture whereas project / CR are getting more into details and this would be a logical info hierarchy ordering to go from high-level to more detailed.

**Review by Ari-S-123 — COMMENTED** [review]
~Date Picker is screwed when clicking, don't merge yet.~

File: src/frontend/src/pages/GanttPage/GanttPageFilter.tsx [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    <PageBlock title="Filters" style={{ flexGrow: 1 }}>
+      <Grid container rowSpacing={1} columnSpacing={1} sx={{ justifyContent: 'start', alignItems: 'start' }}>
+        {carFilters}
+        <Grid item xs={12} md={2}>
+          <FormControl sx={{ width: '100%' }}>
+            <FormLabel>Status</FormLabel>
+            <Select value={status} onChange={statusHandler}>
+              <MenuItem value="All Statuses">All Statuses</MenuItem>
+              {Object.values(WbsElementStatus).map((status) => {
+                return <MenuItem value={status}>{status}</MenuItem>;
```
Comment (jonathan-chen10) [inline src/frontend/src/pages/GanttPage/GanttPageFilter.tsx:142]: Each child in a list should have a unique "key" prop. (appears at least again on line 152 and maybe somewhere else) 
ik its pretty insignificant but i just did the ticket to fix these little bugs and might as well do it now

File: src/frontend/src/pages/GanttPage/GanttPackage/README.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Note
+
+We stole this from here: https://github.com/MaTeMaTuK/gantt-task-react
```
Comment (jonathan-chen10) [inline src/frontend/src/pages/GanttPage/GanttPackage/README.md:3]: lol what if we worded this slightly less sus

**Review by jonathan-chen10 — CHANGES_REQUESTED** [review]
lol this is really something else huh... playing around with it it looks pretty good. i find it a bit strange that we're using a new package instead of figuring out how to reuse/reorganize the project page gantt like a good OOD student would but fair enough we got something to work regardless

maybe we want COMPLETE projects/wps to be a different color from 100% ACTIVE ones or something? right now theyre both the same shade of dark i think

really good work youve been doing! might look over it more carefully some point later on

File: src/frontend/src/pages/GanttPage/GanttPageFilter.tsx [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+          }}
+          startIcon={<UnfoldMoreIcon />}
+        >
+          Expand
+        </Button>
+      </Grid>
+      <Grid item>
+        <Button
+          onClick={() => {
+            expandedHandler(false);
```
Comment (jonathan-chen10) [inline src/frontend/src/pages/GanttPage/GanttPageFilter.tsx:117]: small bug here but it looks like when expanded=false the collapse (and reset?) button doesnt work and when expanded=true the expand button doesnt work. So what that means is if you manually expand or close an individual project gantt these buttons don't undo it. (Try with the default gantt opening a project and clicking collapse/reset) idk if there's a way to force a refresh or something

File: src/frontend/src/pages/GanttPage/GanttPackage/README.md (hunk 2 above)
Comment (anthonybernardi) [reply src/frontend/src/pages/GanttPage/GanttPackage/README.md:3]: commandeered

File: src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx [hunk 4]
```diff
@@ -139,6 +142,7 @@ const GanttPageWrapper: FC = () => {
   const expandedHandler = (value: boolean) => {
     const ganttFilters: GanttFilters = { ...defaultGanttFilters, expanded: value };
     history.push(`${history.location.pathname + buildGanttSearchParams(ganttFilters)}`);
+    window.location.reload(); // quick fix for gantt chart not updating when any expander arrow is clicked
```
Comment (jonathan-chen10) [inline src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx:145]: Cool fix, hopefully people won't mind the page reloading. I think this works for expand/collapse but not for reset.

**Review by jonathan-chen10 — COMMENTED** [review]
Nice fixes!

I'm not too certain on the refresh mechanics of this component. Looks like the manually expanding project thing gets undone if you switch focus to another tab/window and switch back as well. Dunno if there's a way to address that though.

File: src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx (hunk 4 above)
Comment (Ari-S-123) [reply src/frontend/src/pages/GanttPage/GanttPageWrapper.tsx]: Try the next commit

**Review by anthonybernardi — APPROVED** [review]
beautiful

**Pull request merged**